[2단계 - 페이먼츠] 제나(이정은) 미션 제출합니다.

JeLee-river opened the pull request:
안녕하세요 도밥! 주말 잘보내셨나요? ㅎㅎ 이번 2단계 미션도 잘부탁드립니다!

---

## 🎯 페이먼츠

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계
- 재사용 가능한 Input Component를 개발한다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 테스트한다.
- 카드 정보를 효과적으로 렌더링 하기 위한 상태 관리를 경험한다.

### 2단계
- 다양한 Form 구성 요소들간의 상태를 효율적으로 관리한다.
- hooks API를 이용하여 상태 관리 로직을 구현하한다.
- custom hooks를 생성하여 Form 관리 로직을 컴포넌트에서 분리하고 재사용한다.
- Controlled & Uncontrolled Components에 입각하여 Form을 핸들링한다.

---

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [ ] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항(코드 컨벤션, 에러 핸들링 등)을 준수했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 서비스: [링크](https://jelee-river.github.io/react-payments/)
  - storybook: [링크](https://6800ba1c9f8c4a9060a108bb-ozfyptcfmp.chromatic.com/)

<br>

## 코드 의도

### 1. Context API 사용

props drilling과 페이지 단위의 상태 공유 문제 등을 이유로, 사용자에게 입력받는 카드 정보 상태들을 Context API로 관리하도록 리팩토링했습니다.

2단계 미션에서 새로운 페이지와 상태가 추가되면서 카드 정보 state를 공유하는 컴포넌트의 범위가 더 넓어지게 되었습니다. 특히, 두 페이지 모두에 카드 번호와 카드사 정보가 필요했고, 이에 따라 App 컴포넌트에서 상태를 관리해야하는 상황에 이르렀습니다. 그러나 App에서 관리한다면 상태를 사용하는 컴포넌트 간 거리가 4에 달하였습니다. 따라서 컴포넌트의 깊이에 구애받지 않고 상태를 공유할 수 있는 Context API를 도입하여 카드 정보를 관리하기로 했습니다.

<br>

### 2. Select 컴포넌트

디자인 적용이 제한적이라는 단점 때문에 HTML 기본 태그인 Select와 Option 대신 ul, li 를 이용하여 커스텀 Select 컴포넌트를 구현하게 되었습니다.

ul, li를 사용한 이유는 options 역할을 수행하는 엘리먼트가 목록에 가깝다고 생각했기 때문이고, input 태그를 select 태그처럼 사용한 이유는 상위 컴포넌트인 form 으로 관리하기 위해서 입니다.


<br>


## 리뷰 요청 & 논의하고 싶은 내용

#### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

이번 미션에서 가장 많이 고민했던 문제는 미션 전반적인 관점에서는 커스텀 훅의 분리를, 지엽적인 측면에서는 UI의 순차적 렌더링 방식에 대해 가장 많이 고민했습니다. 현재 저는 커스텀 훅의 분리 기준을 'react hook을 사용하는 반복되는 로직'으로 설정했습니다. 따라서 유효성 상태(isValid)와 같이 여러 곳에서 사용되면서, 어느 곳에 두어도 결합도가 낮다고 생각한 로직은 커스텀 훅으로 분리했습니다. 

두 번째로는 UI의 순차적 렌더링 방식에 대해 고민했습니다. 저의 경우, 순차적으로 렌더링 되어야 하는 컴포넌트를 인자로 받아, 렌더링를 수행하는 범용적인 함수를 만들고자 했습니다. 일반적으로 사용되는 step 방식이나 `&&` 연산자를 사용한 조건부 렌더링 방식이 다소 복잡하다고 느꼈기 때문입니다. 또한 컴포넌트에 상관 없이 렌더링을 수행하는 함수를 만들고 싶은 목표도 있었습니다. 그러나 타입 추론 문제 등으로 어려움이 있었고 현재 1차 제출한 코드에는 타입 에러 무시가 적용된 상태입니다. 최종 머지 되기 전까지 이를 개선하거나 다른 방식을 좀 더 고민해보고자 합니다.


#### 2) 이번 리뷰를 통해 논의하고 싶은 부분

### 1. 공통 컴포넌트가 다른 컴포넌트를 의존하는 구조

공통(commom) 컴포넌트가 또 다른 컴포넌트에 의존하는 구조가 적절한지 고민이 있었습니다.

예를 들어, Select 컴포넌트를 구현할 때, 내부의 <Input> 요소를 기본 HTML 태그 그대로 사용할지 Input 컴포넌트를 활용할지 고민했습니다. 하지만 Select 컴포넌트는 공통 컴포넌트로 설계하였기 때문에 최대한 다른 컴포넌트의 의존성를 줄이고자 기본 Input 태그를 사용하기로 했습니다.

이처럼 공통 컴포넌트가 또 다른 공통 컴포넌트를 참조하는 패턴이 일반적인지, 어색하지는 않을지 궁금합니다!

<br>

### 2. Context API 상태 관리 범위

Context API로 꼭 필요한 데이터만 관리해도 될지, 아니면 응집성과 일관성을 위해 같은 범주의 모든 상태를 함께 관리하는 것이 좋을지 고민했습니다.

저의 경우, 처음에는 페이지 간 공유가 필요한 카드 번호와 카드사 정보만 Context API로 관리했습니다. 이후 모든 카드 관련 정보가 두 페이지 사이에서 공유되어야 하면서 자연스럽게 카드와 연관된 모든 상태를 Context API로 옮겼습니다. 
하지만 만일 추가로 공유할 데이터가 없었더라면, 처음처럼 카드 번호와 카드사 정보만 Context API로 유지했을 것입니다. 다른 상태도 모두 Context API로 관리해야할 필요성을 느끼지 못했기 때문입니다. 그러나 이렇게 일부 데이터만 다른 방식으로 상태를 관리하는 것이 일관성 면에서 떨어지지는 않을지 우려가 됩니다. 

이처럼 같은 범주에 속한 상태 중 일부만 Context API로 관리할 때, 코드를 읽는 사람이 혼란을 느끼거나 통일성이 떨어진다고 보시나요? 도밥의 의견이 궁금합니다.

<br>

### 3. 테스트를 위한 class 또는 data- 속성 추가

테스트를 위해 class나 data- 속성을 추가하는 방식에 대해 어떻게 생각하시는지 궁금합니다.

스토리북으로 컴포넌트를 테스트할 때, 특정 <Input>에 접근하기 위해 label이나 placeholder 같은 식별자가 필요했습니다. 하지만 해당 속성이 없는 경우, 테스트 코드에서 원하는 엘리먼트를 찾아내기 어려웠습니다. 

이 문제를 해결하기 위해 테스트 전용으로 class나 data- 속성을 추가하는 방안을 고민했으나,
서비스 코드에 테스트에만 사용하는 코드를 추가하는 것이 적절한지 의문이 들었습니다.

결국 input의 name에 접근하는 방식을 찾아 이 고민은 해결했지만, 테스트에만 사용되는 class/data- 속성을 실제 코드에 추가하는 방식에 대해 어떻게 생각하시는지 궁금합니다.

<br>



---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도만이 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. Form 상태 관리 & Custom Hook 분리

- 반복되는 로직을 custom hook으로 분리했는가?
- hook 내부와 UI 컴포넌트의 **역할이 명확하게 분리**되어 있는가?
- 상태 흐름이 직관적이며, Form 전체를 일관되게 관리할 수 있는 구조인가?

### 2. 입력 UI 흐름과 UX

- 카드 번호 입력 시 필드 간 **자동 포커싱 이동**이 자연스럽게 동작하는가?
- 숫자만 입력 가능한 필드에서 제한, 에러 메시지, 유효성 피드백 등 **사용자 경험이 충분히 고려**되었는가?

### 3. 컴포넌트 구조 및 재사용성

- 컴포넌트가 명확한 **역할과 책임**을 가지며 과도하게 분리되거나 중첩되어 있지 않은가?
- Input, Button 등 재사용 가능한 컴포넌트가 잘 정의되어 있는가?

### 4. 상태 기반 유효성 검사 및 확인 버튼 활성화

- 모든 필드가 유효할 때만 확인 버튼이 **정확히 활성화/비활성화**되는가?
- 유효성 검사의 기준이 명확하고, **상태 변경에 따른 UI 반응**이 잘 연결되어 있는가?

---

감사합니다!

wmakerjun committed:
docs: PR 템플릿 추가

JeLee-river committed:
refactor: 클래스명의 불필요한 prefix 제거

JeLee-river committed:
fix: input 컴포넌트의 스토리에 css 및 컴포넌트 import 경로가 잘못된 문제 수정

JeLee-river committed:
fix: 두번째 카드번호부터 입력할때 로고가 렌더링되는 문제 개선

- 카드번호 배열을 문자열로 파싱하면서, 두번째 번호가 첫번째 처럼 변경되는 문제를 수정함

- carNumbers에 초기값을 부여하여, 입력값을 받기 전에도 프리뷰가 정상적으로 렌더링될 수 있도록 수정

JeLee-river committed:
refactor: 컴포넌트 Input이 input 태그의 일부 속성만 가져오도록 리팩토링

JeLee-river committed:
refactor: 로고 이미지 태그에 alt 속성 추가

JeLee-river committed:
chore: storybook 배포를 위한 명령어 추가

JeLee-river committed:
chore: 2단계 요구사항에 따라 react-router 설치

JeLee-river committed:
docs: 추가 기능 요구 사항을 반영하여 기능 목록 업데이트

JeLee-river committed:
feat: 드롭박스 형식의 Select 컴포넌트 생성

JeLee-river committed:
style: 컴포넌트 Select에 hover했을때 커서 스타일 변경

JeLee-river committed:
feat: 컴포넌트 SelectBox에 드롭다운 아이콘 추가

JeLee-river committed:
feat: 카드사를 선택하는 InputForm을 추가

JeLee-river committed:
style: 카드사 선택 UI의 옵션들이 숨김 처리될때 레이아웃을 유지하지 않도록 수정

JeLee-river committed:
refactor: 컴포넌트 Select를 제어 컴포넌트로 설정하여 state를 props로 받도록 리팩토링

JeLee-river committed:
feat: 사용자가 선택한 카드사에 따라 프리뷰 색상이 변경되는 기능 추가

JeLee-river committed:
refactor: 통일성을 고려하여 일관된 컨벤션을 유지하도록 불필요한 코드 수정

JeLee-river committed:
refactor: 유효기간과 카드번호 컴포넌트가 유효성 상태를 상위 컴포넌트에서 받도록 수정

JeLee-river committed:
refactor: props의 타입을 별도로 정의하도록 리팩토링

JeLee-river committed:
chore: 변수명 오타 수정

JeLee-river committed:
refactor: 컴포넌트 CardCVCInput이 cvc 상태를 상위 컴포넌트에서 받도록 수정

JeLee-river committed:
refactor: 불필요한 aria- 타입 제거

JeLee-river committed:
refactor: 카드 번호 input에 key 값 부여

JeLee-river committed:
docs: 잘못 작성된 요구사항 수정 및 누락된 요구사항 추가

JeLee-river committed:
docs: 완료한 요구 사항 업데이트

JeLee-river committed:
feat: 카드 비밀번호 컴포넌트 생성

JeLee-river committed:
fix: 카드 번호 상태가 바뀌지 않는 문제 수정

wmakerjun committed:
docs: PR 템플릿에 2단계 학습 목표 및 리뷰어 가이드 추가

JeLee-river committed:
feat: 사용자가 입력을 완료하면 focus가 다음 input으로 이동하는 기능 구현

JeLee-river committed:
test: 컴포넌트 CarPreview에 cardIssuer가 props로 추가됨에 따라 args 추가

JeLee-river committed:
feat: 컴포넌트 Button 구현

JeLee-river committed:
feat: 카드 정보 폼의 확인 버튼 구현

JeLee-river committed:
chore: 기존의 react-router 대신 웹 개발 기능을 추가 제공하는 react-router-dom 설치

JeLee-river committed:
refactor: 프로젝트 전반에 적용되는 css 파일들을 global 디렉터리로 이동

JeLee-river committed:
feat: 모든 폼 데이터가 유효할때 확인 버튼이 생성되는 기능 추가

JeLee-river committed:
test: 컴포넌트들이 isValid를 props로 전달받게 되면서 test 코드 수정

JeLee-river committed:
feat: 카드 정보 입력 페이지에 라우터 적용

JeLee-river committed:
feat: 입력값이 유효할때 확인 버튼으로 등록 완료 페이지에 접근하도록 라우터 추가

JeLee-river committed:
feat: registerComplete 페이지 컴포넌트 추가

JeLee-river committed:
style: 등록 완료 페이지 스타일 적용

JeLee-river committed:
style: 카드사 선택 컴포넌트 스타일 수정

JeLee-river committed:
refactor: 카드번호, 카드사 상태를 ContextAPI로 관리하도록 리팩토링

- 카드번호, 카드사 상태는 카드 정보 입력 페이지와 카드 등록 완료 페이지 모두 공유함

- 이를 최상위 컴포넌트 App에서 props를 내려주면 Drilling 깊이가 4임

- 이에 따라 ContextAPI를 사용하는 것이 깔끔하다고 판단함

JeLee-river committed:
test: props 변경에 따른 CardNumberInput 컴포넌트 테스트 수정

JeLee-river committed:
chore: 불필요한 세미콜론 제거

JeLee-river committed:
refactor: 등록 완료 문구 수정

JeLee-river committed:
test: 컴포넌트 CardIssuerSelector의 테스트 추가

JeLee-river committed:
test: 컴포넌트 변경에 따른 테스트 코드 수정

JeLee-river committed:
chore: 불필요한 코드 제거

JeLee-river committed:
feat: 카드 정보 폼의 유효성 검사를 관리하는 커스텀 hook 생성

JeLee-river committed:
refactor: 카드 정보 입력 input에 required 속성 부여

JeLee-river committed:
refactor: input의 클래스명이 자연스럽게 되도록 조건부 클래스의 순서 변경

JeLee-river committed:
refactor: 통일성 및 가독성을 고려하여 변수명 및 함수명의 expiration을 expiry로 수정

JeLee-river committed:
refactor: 카드 유효기간 예외 처리 순서 변경

JeLee-river committed:
refactor: 컴포넌트 Input에 최소 입력 길이 속성 추가

JeLee-river committed:
test: 컴포넌트 Input의 테스트 오류 수정

JeLee-river committed:
refactor: 유효기간 예외 처리 로직 수정

JeLee-river committed:
refactor: 유효기간 유효성을 검증하는 로직을 validate 함수로 분리

- 유효기간 컴포넌트의 유효기간 검증 로직을 별도의 파일로 분리함

- 유효기간이 모두 입력되었을때 검증을 수행하도록 변경함

- 상태변경이 비동기적으로 수행되는 점을 고려하여 키보드 이벤트가 발생할 때, 변경되는 항목은 state 대신 사용자 입력값을 바로 사용하도록 수정함

JeLee-river committed:
refactor: 카드 유효기간 검증 로직 수정

- 불필요한 isValid 값을 제거하고 기존의 fields를 isValid에 할당하도록 수정

- 년도 검증 로직을 제거하고 전체 유효기간을 검증하도록 수정

JeLee-river committed:
refactor: 사용자가 숫자가 아닌 값을 입력하면 숫자와 관련된 메시지가 우선 출력되도록 수정

JeLee-river committed:
fix: 카드 유효기간이 유효하지 않을때 피드백 메시지가 생성되지 않는 문제 수정

- 숫자 체크 함수로 인에 feedbackMessage가 덮어씌여지고 있었음

JeLee-river committed:
test: storybook에 피드백 메시지가 나타나지 않는 문제 수정

JeLee-river committed:
chore: 불필요한 코드 제거

JeLee-river committed:
feat: 비밀번호 컴포넌트를 폼에 추가

JeLee-river committed:
refactor: form의 유효성을 동적으로 체크하고 버튼을 나타나게 하도록 리팩토링

JeLee-river committed:
style: select 박스의 option이 전체 레이아웃을 변형하며 열리지 않도록 수정

JeLee-river committed:
docs: 구현한 기능 목록 정리

JeLee-river committed:
test: 컴포넌트 CardPasswordInput의 스토리 작성

JeLee-river committed:
test: 로직이 정상임에도 카드 CVC 컴포넌트 테스트가 통과되지 않는 문제 수정

JeLee-river committed:
fix: 등록 완료 이미지가 보이지 않는 문제 수정

JeLee-river committed:
style: 카드 등록 완료 페이지 스타일 조정

JeLee-river committed:
refactor: 모든 카드 정보 상태를 Context API로 관리하도록 리팩토링

- 모든 카드 컴포넌트들이 제어 컴포넌트로 관리되고 있으므로 폼을 제출할 때 직접 초기화해야함

- 폼에서 상태에 접근하기 위해 state를 전달해줘야함

- 남은 state들도 Context로 관리하여 코드를 간결화하도록 결정

JeLee-river committed:
refactor: 카드 정보 입력 페이지로 돌아갈때 폼이 초기화되도록 수정

JeLee-river committed:
refactor: 디렉터리명과 중복되는 Card prefix를 제거

- 불필요하게 네이밍이 길어진 것을 수정

JeLee-river committed:
test: 상태관리 방식 변경에 따른 테스트 코드 수정

JeLee-river committed:
feat: 컴포넌트 렌더링을 관리하는 커스텀 훅 생성

JeLee-river committed:
refactor: select 컴포넌트가 내부 input의 유효성을 검증하도록 리팩토링

JeLee-river committed:
refactor: selector가 유효성을 검증하는 로직 수정

JeLee-river committed:
refactor: 렌더링 관리 커스텀 훅 로직 수정

JeLee-river committed:
refactor: input이 빈 문자열이 아닌 경우 isNotValid 클래스를 추가하도록 수정

JeLee-river committed:
refactor: any를 사용하지 않도록 타입 수정

JeLee-river committed:
style: 카드 등록 페이지 스타일 조정

JeLee-river committed:
Merge branch 'woowacourse:main' into step2

JeLee-river committed:
fix: 배포 페이지 경로를 찾을 수 없는 문제 수정

JeLee-river committed:
chore: vite 설정파일에 배포 경로 추가

JeLee-river committed:
fix: 경로 환경변수가 잘못 참조된 문제 수정

JeLee-river committed:
fix: 등록 완료 사진이 표시되지 않는 문제 수정

JeLee-river committed:
fix: 배포 페이지 경로 수정

JeLee-river committed:
fix: 카드사, 드롭다운 이미디가 보이지 않는 문제 수정

JeLee-river commented:
안녕하세요 도밥!
현재 UI를 순차적으로 표시하는 로직에서 타입 관련 문제가 존재합니다. 이를 기한까지 수정하는 것에 어려움이 있을 것 같아, 우선 리뷰 요청을 먼저 드리고자 해당 상태 그대로 제출하였습니다. 😭

또한 UI가 표시 과정에서 확인 버튼이 조건을 충족하지 않았음에도 노출되는 문제가 있습니다. 도밥의 1차 피드백을 반영하며 이 부분도 함께 수정하도록 하겠습니다.

감사합니다!

Creative-Lee commented on src/hooks/useCard.ts:
```diff
@@ -0,0 +1,13 @@
+import { useContext } from 'react';
+import { CardContext } from '../contexts/CardContext';
+
+function useCard() {
```
[nit] 컨텍스트의 value를 return하는 랩핑함수의 네이밍에는 
보통 끝에 `Context`를  붙이는것 같아요. (e.g. useCardContext)

JeLee-river replied:
Context api를 관리하는 훅의 이름을 변경하였습니다!

[refactor: useCard 커스텀훅명에 Context suffix 추가 (eef3990)](https://github.com/woowacourse/react-payments/pull/473/commits/eef3990aab977e84c72a3adf4be264a6ef56be95)

Creative-Lee commented on src/hooks/useCard.ts:
[수정 요청 아님]

지금은 서비스 규모가 작아 hook, contexts 등의 디렉터리로 각각 나누어 두어도 좋은 것 같아요.
`context+customHook`와 같이 자연스레 같은 관심사를 가진 코드의 경우, 둘을 합쳐 하나의 파일에 작성하기도 합니다. 규모가 커졌을때 관리하기 좋기도 하고요

JeLee-river replied:
오 오히려 서비스가 커졌을 때 관리 편의성을 위해 코드를 합친다는 관점은 생각해보지 못했습니다. 규모 있는 서비스에서는 코드 분리만 중요할 것이라 생각했는데, 관심사를 고려하여 하나로 합치는 것이 유지보수에 더 좋을 수 있다는 점을 인지하게 되었습니다. 👍  좋은 코멘트 감사합니다!

Creative-Lee commented on src/hooks/useValidateForm.tsx:
```diff
@@ -0,0 +1,57 @@
+      isCVCValid &&
+      isPasswordValid;
+
+    setIsDataValid(isTotalDataValid);
```
`isTotalDataValid` 라는 변수 자체가 상태에서 파생된 파생상태(계산된 상태) 입니다.
불필요한 state(isDataValid)가 만들어진 상황이에요.

[react 공식문서 -  avoid-redundant-state](https://ko.react.dev/learn/choosing-the-state-structure#avoid-redundant-state) 를 첨부할테니 해당 섹션을 꼭 읽어보시고 불필요한 상태를 제거해봅시다.

Creative-Lee replied:
더불어 이 useEffect는 올바른 사용사례가 아닙니다. 
[react 공식문서 - you-might-not-need-an-effect ](https://ko.react.dev/learn/you-might-not-need-an-effect) 도 첨부할테니 꼭 정독해보시고 useEffect를 어떤 경우에 사용해야하는지 확실하게 학습해보도록 합시다💪

Creative-Lee replied:
이번 미션에서 `useState`, `useEffect` hook의 베스트 유즈 케이스만 배워가도 과장 조금 보태서 react 다 배웠다 싶을 정도로 진심으로 중요합니다. 반드시 학습해봅시다.

JeLee-river replied:
공식 문서 예시에 나온 잘못된 사용 방식을 제가 그대로 따라하고 있었네요. 😂 

useEffect를 단순히 '참조값이 바뀐 뒤에 실행되는 훅'으로만 생각하고, 상태 변경 후 실행해야 할 로직까지 모두 사이드 이펙트로 잘못 간주하고 사용했습니다. 그러나 실제로는  그 로직들이 React의 렌더링 흐름만으로 충분히 처리될 수 있고, useEffect는 외부와의 상호작용과 같은 진짜 사이드 이펙트에만 사용해야 한다는 점을 놓치고 있었습니다. 
또한, 상태로 관리되는 값을 참조하고 있으면, 상태 변경 시 자동으로 업데이트된다는 사실도 인지하지 못하고 불필요한 상태를 생성하고 있었습니다.

도밥이 공유해 주신 자료 덕분에 무엇이 잘못되었는지 이해하고 개선할 수 있었습니다. hook 사용법을 겉핥기 식으로만 알고 실제 적용 방식은 분명히 알지 못한 점도 깨달았습니다. 정말 감사합니다. 👍 

[refactor: useValidateForm의 불필요한 useEffect 제거 (6d96dac)](https://github.com/woowacourse/react-payments/pull/473/commits/6d96daccab832120e38bc0c553cb25a95a106a1f)

Creative-Lee replied:
중요한 부분인데 잘 학습하신 것 같아 다행입니다👍

Creative-Lee commented on src/hooks/useValidateForm.tsx:
```diff
@@ -0,0 +1,57 @@
+    true,
+    true,
+  ]);
+  const [isIssuerValid, setIsIssuerValid] = useState(false);
```
유효한지 여부가 처음부터 true인것과 false인것은 어떤 차이가 있나요? 
훅 로직파악에 혼동이 생길 수 있는 부분인 것 같습니다.

JeLee-river replied:
전자의 경우, 초기 상태에서 아무 입력도 없는 것은 아직 검증 전이라고 보고, boolean 기본값을 true로 두는 것이 더 자연스럽다고 생각했습니다.
하지만 후자인 카드사의 경우, 선택하지 않은 것 자체를 '유효하지 않음'으로 간주하기 때문에, 기본값을 false로 설정하는 것이 적절하다고 판단했습니다.

그러나 도밥의 말씀처럼, 결국 동일한 카드 정보임에도 기준이 다른 점이 확실히 어색하게 느껴지는 것 같습니다. 😂 
카드사를 선택하지 않은 상태를 true로 수정하는 것보다, 입력이 없는 상태를 false로 수정하여 통일하는 것이 논리상 좀 더 자연스러운 것 같아, 모든 카드 정보의 기본값을 false로 수정하겠습니다!

JeLee-river replied:
카드 정보의 유효성 state 기본값을 false로 통일했습니다!

[refactor: 카드 정보의 유효성 state 기본값을 false로 통일 (7032186)](https://github.com/woowacourse/react-payments/pull/473/commits/7032186f18f98eba42f36831ee81a93ea730c425)

Creative-Lee commented on src/utils/precise.ts:
```diff
@@ -1,4 +1,4 @@
-export const validatorUtils = {
+export const precise = {
```
[nit] 어떤 의미로 변경된 파일명과 변수명이 변경된걸까요~?

JeLee-river replied:
validatorUtils 객체로 묶인 함수들은 유효성 검증 자체보다는 유효성 기준에 맞는지 판단만 하고 있다고 생각하여, precise로 바꾸었습니다. 그리고 이를 이용하여 도메인에 맞는 형태(isValid, message)로 결과를 반환하는 로직은 validate 파일에 구현했습니다!

Creative-Lee commented on src/utils/validate.ts:
```diff
@@ -0,0 +1,43 @@
+        isValid: { month: true, year: true },
+        message: '',
+      };
+    }
```
month, year 둘중 하나라도 2자리 씩 입력이 되지 않았다면, 유효하지 않은 상태인데
로직에 오류가 있는 것 같아요.

JeLee-river replied:
전반적으로 유효기간을 체크하는 로직을 개선하여, month와 year의 유효성 검사가 독립적으로 수행되도록 변경했습니다!
1. 하나로 통일되었던 month와 year의 onChange 핸들러 함수를 두 개로 분리했습니다. 이 과정에서 코멘트 주신, 길이를 비교하던 조건문을 제거했습니다. 

2. 1번에서 추가된 핸들러 함수 내부에서 검증 결과가 유효하지 않을 경우에만 isValid 상태가 업데이트되던 문제를 개선했습니다.

3. 사용자가 입력을 완전히 마친 후, 입력값을 검증하고 메시지를 출력하도록 하고자, 각 핸들러 함수에 input value의 길이를 체크하고 유효성을 검사하도록 코드를 수정했습니다. 


[refactor: 유효기간의 월과 년도에 개별적으로 적용되도록 유효성 검증 로직 수정 (729b9eb)](https://github.com/woowacourse/react-payments/pull/473/commits/729b9eb599b4654a78468c0fe15cd4310f0139c3)

[refactor: 카드 유효기간의 유효성 검사 결과가 true일 경우에도 상태를 업데이트하도록 수정 (561d3b4)](https://github.com/woowacourse/react-payments/pull/473/commits/561d3b4b122d1ab7fabf50dd804ed07ac9aafe1a)

[refactor: 두 필드의 길이가 2일때 만료일을 검증하도록 수정 (79e430c)](https://github.com/woowacourse/react-payments/pull/473/commits/79e430c3a2e227dd7a26ce0f2866ca2e6aec3423)

Creative-Lee commented on src/utils/validate.ts:
```diff
@@ -0,0 +1,43 @@
+        isValid: { month: false, year: true },
+        message: '유효하지 않은 월 형식입니다. 유효 기간을 확인해주세요.',
+      };
+    }
```
year가 유효하지 않았더라도, month가 유효하지 않다면 year가 유효해지는 상황이 생길 것 같아요. 
실제로 동작에 오류가 없더라도, 논리적으로는 잘못된 코드입니다.
다른 방법으로 문제를 해결해야할 것 같아요.(month와 year의 유효성을 따로 검사하는 등)

JeLee-river replied:
동일한 onChange 핸들러 함수를 적용했던 기존의 코드에서, 핸들러 함수를 분리하여 월과 년도에 개별적으로 적용되도록 하고, 현재 입력 중인 input의 유효성 상태만 변경되도록 수정했습니다!

[refactor: 유효기간의 월과 년도에 개별적으로 적용되도록 유효성 검증 로직 수정 (729b9eb)](https://github.com/woowacourse/react-payments/pull/473/commits/729b9eb599b4654a78468c0fe15cd4310f0139c3)

Creative-Lee commented on src/utils/validate.ts:
```diff
@@ -0,0 +1,43 @@
+      };
+    }
+
+    if (!precise.hasNotExpired({ month, year })) {
```
부정을 다시 부정하는 것은 코드 파악 과정에 논리적으로 혼란을 주기에 
가능한 긍정으로 변경하는 게 좋습니다. 

as-is: 만료되지 않은것의 반대
to-be: 만료됨

JeLee-river replied:
확실히, 오류를 해결하기 위해 해당 로직을 살펴볼때 헷갈려서 불편함을 느꼈는데, 부정의 부정보다는 긍정이 훨씬 이해하기 쉽고 좋은 것 같습니다. 코멘트 감사합니다! 👍

JeLee-river replied:
`hasNotExpired`를 `isExpiredDate`로 변경하였습니다!

[refactor: 논리 파악에 혼란을 주기 쉬운 예외 처리 문법 개선 (378e0f0)](https://github.com/woowacourse/react-payments/pull/473/commits/378e0f0da4f67ed47f5d5f72693c0b8e39cb1c8f)

Creative-Lee commented on src/hooks/useRender.tsx:
```diff
@@ -0,0 +1,46 @@
+  renderItems: RenderItem[],
+  ref: React.RefObject<HTMLFormElement>
+) {
+  const [renderList, setRenderList] = useState<RenderItem[]>([
+    renderItems[START_INDEX],
+  ]);
```
`RenderItem []` type을 가지는 `renderItems 인자`를 받아서 
useState에 초기값에 전달할 때 `renderItems[0]`을 할당한다면 
이 값의 type은 `RenderItem` 입니다.

그렇다면 state 이름도 `renderList`가 아니지 않을까요?

JeLee-river replied:
오 제가 같은 state를 지칭하면서 `renderItems`와 `renderList`라는 서로 다른 변수명을 사용했었네요..! 의미상 동일한 변수는 하나의 이름 규칙으로 통일해야 하는 점을 놓쳤습니다. 좋은 코멘트 감사합니다!

JeLee-river replied:
카드 유효기간의 유효성 상태(isValid) 변경값이 컴포넌트에 반영되지 않던 문제를 해결하면서 해당 코드가 제거되었습니다!

[fix: 카드 유효기간 유효성 검사 결과 변경 시 상태가 갱신되지 않던 문제 해결 (defd61e)](https://github.com/woowacourse/react-payments/pull/473/commits/defd61e806f2241d971589036fdb883b84254840)

Creative-Lee commented on src/hooks/useRender.tsx:
useValidateForm.ts에 남긴 코멘트가 동일하게 적용됩니다.
아래 코멘트를 먼저 읽고 동일하게 훅을 개선해봅시다.

JeLee-river replied:
useEffect를 제거해도 의도했던 동작이 그대로 수행되는 것을 확인하고, 제가 `useEffect`를 확실하게 이해하지 않고 사용했다는 점을 깨달았습니다.  상태를 변경할때 불필요하게 `useEffect`를 사용하고 있었네요. 😂 좋은 코멘트 감사합니다!!

[refactor: useRender의 불필요한 useEffect 제거 (2486c59)](https://github.com/woowacourse/react-payments/pull/473/commits/2486c59275130d0bfc2a2f70fc83bc71a632284e)

Creative-Lee replied:
훌륭합니다👍

Creative-Lee reviewed (changes requested):
안녕하세요 제나. 다시 만나서 반가워요😄
저도 잘 부탁드립니다 ㅎㅎ

---
질문해주신 내용을 살펴보면, 

> 1️⃣ 공통 컴포넌트가 다른 컴포넌트를 의존하는 구조
공통(commom) 컴포넌트가 또 다른 컴포넌트에 의존하는 구조가 적절한지 고민이 있었습니다.
(... 생략) 하지만 Select 컴포넌트는 공통 컴포넌트로 설계하였기 때문에 최대한 다른 컴포넌트의 의존성를 줄이고자 기본 Input 태그를 사용하기로 했습니다.

좋은 고민 끝에 잘 선택해주셨다고 생각해요 👍 
말씀해주신 것처럼 `Input`이 `Select` 보다 더 원시적인 성향을 가졌다고 볼 수 있는데, Input 컴포넌트가 어떻게 정의 되느냐에 따라 Select 구현이 달라질 수 있겠죠.(Select의 구현을 위해 Input을 수정하게 되버리는 이상한 상황이 생길수도 있을 것 같고요. 계층이 많아 질수록 생각해야할 부분이 많아질것 같아요.)

Input과 같은 아주 작은 원시레벨의 컴포넌트를 아예 본적이 없는 것은 아니지만, 
개인적으로는 위와 같은 이유로 InputFiled, Select, Radio 와 같은 규모의 컴포넌트를 base로 하는 공용 컴포넌트 시스템을 선호하고, 이에 익숙한것 같습니다. 

결론
- `Input 원시 컴포넌트`로  InputFiled, SelectBox, Radio 같은 컴포넌트를 구현하는 계층구조(1 : N)는 익숙하지 않고 거의 본적이 없다. 
- Radio 정도의 컴포넌트를 RadioGroup같은 컴포넌트로 확장하여 사용하는 케이스(1 : 1)는 많이 보았다.


> 2️⃣ Context API 상태 관리 범위
Context API로 꼭 필요한 데이터만 관리해도 될지, 아니면 응집성과 일관성을 위해 같은 범주의 모든 상태를 함께 관리하는 것이 좋을지 고민했습니다.

딜레마가 있을 수 밖에 없는 부분에서 좋은 고민을 해주셨네요👍 
지금 단계에서 두 방식 중,`뭐가 옳다. 정답이 있다.` 라고 볼 문제는 아닌 것 같아요ㅎㅎ
둘다 합당한 구현이라고 생각합니다. 

> 저의 경우, 처음에는 페이지 간 공유가 필요한 카드 번호와 카드사 정보만 Context API로 관리했습니다. 이후 모든 카드 관련 정보가 두 페이지 사이에서 공유되어야 하면서 자연스럽게 카드와 연관된 모든 상태를 Context API로 옮겼습니다. 

'필요한 상태는 a, b 뿐인데, c, d, e 까지 공유해야해?', '필요한 데이터만 공유하겠어.'라는 판단도 합당하다고 생각하고, 
 
> 그러나 이렇게 일부 데이터만 다른 방식으로 상태를 관리하는 것이 일관성 면에서 떨어지지는 않을지 우려가 됩니다. 

마찬가지로 '같은 맥락(카드 등록)내의 상태인데 a, b, c, d, e 모두 같이 묶어야 하는 것 아니야?', '왜 뭐는 넣고 뭐는 빼?' 라는 우려도 합당하다고 생각해요. 

중요한 것은 제나가 잘 정의해주신 것 처럼 `선택의 기준`(이렇게 생각하기에, 요렇게만 사용하겠어)인데, 이런 부분에서 훌륭하다고 생각합니다👍 

다른 케이스를 제외하고, 다른 페이지에서 form의 모든 데이터가 필요한게 아닌 상황만 놓고 본다면, 개인적으로도 필요한 상태만 context에 담아 전달했을 것 같아요. 

---

> 3️⃣ 테스트를 위한 class 또는 data- 속성 추가
> 서비스 코드에 테스트에만 사용하는 코드를 추가하는 것이 적절한지 의문이 들었습니다.

레벨1 에서 학습이 아주 야무지게 되었군요 제나..! 좋은 고민을 해주셨어요👍
class는 잘 모르겠지만, 테스팅을 위한 `dataset` 추가는 일반적인 방법이라고 알고있어요 ㅎㅎ
관련 내용이 담긴 [RTL의 공식문서](https://testing-library.com/docs/queries/bytestid/)를 첨부할게요~!
(TMI: 빌드 타임에 `test-id`와 같은 테스트용 dataset을 제거해주는 처리도 가능합니다)

---

고생하셨어요 제나~!
다음 미션과 피드백 반영이 겹치게 되었지만, 힘내보시죠 💪

Creative-Lee commented:
요구사항에 대한 부분과 UX 피드백도 조금 드리면,

- LMS 피그마 시안과 동일한 모바일 너비를 가지는 디바이스(iPhone SE)에서 레이아웃 넓어짐

<img src="https://github.com/user-attachments/assets/f4ce30e6-d803-4be1-ad5a-e653b8cd8e66" width="300"/> 

- 제출 후 뒤로가기 가능 + 정보가 남아있는데 동작이 어색함(지우고 다시 쳐야 제출가능함)
<img src="https://github.com/user-attachments/assets/8fb66990-65b1-48fc-9595-362084bbcc4b" width="300" /> 


- form 하위 input 요소들이 피그마 시안과 반대 순서로 추가됨 
  - 피그마: 최신 input이 상단으로
  - 제나: 최신 input이 하단으로 

- cvc 번호 입력 시 카드의 뒷면을 보여주는 처리가 누락됨 (요구 사항에 있었던것 같은데, 혹시 사라진거라면 무시하셔요 ㅎㅎ)
<img width="280" alt="image" src="https://github.com/user-attachments/assets/094d1481-0737-4189-bd82-9abe42cfe808" />

JeLee-river committed:
feat: 각 input의 isValid 값을 컴포넌트 단위로 통합하는 유틸함수 생성

- 컴포넌트 단위로 입력값이 유효한지 검증할때 결과형식을 통일할 수 있도록 유틸함수 생성

JeLee-river committed:
fix: useRender 커스텀 훅 사용시 발생하는 타입 오류 문제 개선

JeLee-river committed:
refactor: useCard 커스텀훅명에 Context suffix 추가

- useCard가 Context API 관련 hook임을 고려하여 Context라는 suffix를 추가

JeLee-river committed:
refactor: 유효기간의 월과 년도에 개별적으로 적용되도록 유효성 검증 로직 수정

JeLee-river committed:
refactor: 논리 파악에 혼란을 주기 쉬운 예외 처리 문법 개선

JeLee-river committed:
refactor: 카드 유효기간의 유효성 검사 결과가 true일 경우에도 상태를 업데이트하도록 수정

JeLee-river committed:
fix: 카드 유효기간 유효성 검사 결과 변경 시 상태가 갱신되지 않던 문제 해결

- 문제: 컴포넌트에서 유효성 상태가 변경되어도 항상 false로 적용되었음

- 원인: renderList에 담기는 컴포넌트 props가 복사되어 담기므로 상태 업데이트가 불가함

-

JeLee-river committed:
refactor: value 객체 분해 형식 통일

JeLee-river committed:
refactor: 파일명과 컴포넌트명을 통일

JeLee-river committed:
chore: 코드 줄바꿈 형식 통일

JeLee-river committed:
refactor: 유효성 상태 변수명이 각 카드 정보 항목을 구체적으로 포함하도록 개선

JeLee-river committed:
refactor: 예외처리 로직에서 precise 직접 호출을 validate 함수 사용으로 대체함

JeLee-river committed:
chore: 불필요한 중복 context provider 제거

JeLee-river committed:
refactor: react가 list items 형태로 렌더링되는 input들을 인식할 수 있도록 key 부여

JeLee-river committed:
refactor: 에러 메시지를 상수로 분리

JeLee-river committed:
refactor: 카드 정보의 유효성 state 기본값을 false로 통일

- 카드 정보의 유효성 state 기본값을 통일하기 위해, 기존에 true로 설정된 state를 false로 변경

JeLee-river committed:
refactor: 두 필드의 길이가 2일때 만료일을 검증하도록 수정

- 사용자가 입력을 완전히 마친 후 유효성을 검증하도록 하기 위해 개선

JeLee-river committed:
refactor: useValidateForm의 불필요한 useEffect 제거

JeLee-river committed:
refactor: useRender의 불필요한 useEffect 제거

JeLee-river committed:
refactor: Input 컴포넌트에 pattern 속성 추가

- 사용자의 입력값의 길이를 엄격하게 검사하기 위해 pattern 속성을 추가

- minLength 검사는 기본적으로 폼을 submit 할때만 제약이 통하도록 설계되는 경우가 많기 때문임.

- 반면, pattern은 입력값이 바뀔때마다 검사함

JeLee-river committed:
refactor: 피그마 요구사항에 맞게 카드 정보 입력 UI가 아래에서 위로 누적되도록 수정

JeLee-river committed:
style: 모바일 환경에서 확인 버튼이 표시되면 레이아웃이 변경되는 문제 개선

JeLee-river committed:
refactor: 카드 정보 state를 초기화하는 함수가 카드 등록 완료 페이지 이동 전에 호출되도록 수정

- 뒤로가기 시, 폼 데이터를 초기화 하기 위해 로직 수정

JeLee-river committed:
refactor: 변수명의 통일성을 고려하여 renderList를 renderItems로 수정

JeLee-river committed:
refactor: 새롭게 렌더링되는 카드 정보 UI에 자동으로 포커싱이 가도록 수정

- 카드 폼을 역순으로 누적하도록 변경하면서, 자동 포커싱 로직을 리팩토링함

JeLee-river commented:
> * LMS 피그마 시안과 동일한 모바일 너비를 가지는 디바이스(iPhone SE)에서 레이아웃 넓어짐

[style: 모바일 환경에서 확인 버튼이 표시되면 레이아웃이 변경되는 문제 개선 (229e8e6)](https://github.com/woowacourse/react-payments/pull/473/commits/229e8e6fab83c206cbbdaca785270f5c2e19b3bd)

<br>

> * form 하위 input 요소들이 피그마 시안과 반대 순서로 추가됨

[refactor: 피그마 요구사항에 맞게 카드 정보 입력 UI가 아래에서 위로 누적되도록 수정 (91f6abf)](https://github.com/woowacourse/react-payments/pull/473/commits/91f6abf0cde3bbf21c51a619b59808098b159129)

<br>

> * cvc 번호 입력 시 카드의 뒷면을 보여주는 처리가 누락됨 (요구 사항에 있었던것 같은데, 혹시 사라진거라면 무시하셔요 ㅎㅎ)

미션 공개 직후에는 존재 했었는데, 이후 요구사항이 업데이트 되면서 사라졌습니다! ㅎㅎ

JeLee-river commented:
안녕하세요, 도밥! 연휴 잘 보내고 계신가요? 😃 피드백 반영 및 수정을 완료하여 리뷰 재요청을 드립니다.

---

## 변경사항

주요 변경 사항은 다음과 같습니다.

<br>

### 1. 유효성 검사 로직 개선

- 월(month)과 년도(year)의 유효성 검사 로직 개선
  - 하나로 통일되었던 month, year의 onChange 핸들러 함수를, 각각의 필드에 독립적으로 적용할 수 있도록 두 개로 분리 [(729b9eb)](https://github.com/woowacourse/react-payments/commit/729b9eb)

  - 두 필드의 길이가 2일 때 만료일을 검증하도록 수정 [(79e430c)](https://github.com/woowacourse/react-payments/commit/79e430c)
 
<br>

### 2. Validation 상태 관리 및 변수명 개선

- 카드 정보의 유효성 state 기본값을 false로 통일 [(7032186)](https://github.com/woowacourse/react-payments/commit/7032186)

- 유효성 상태 변명이 각 카드 정보 항목을 구체적으로 포함하도록 개선 [(069230e)](https://github.com/woowacourse/react-payments/commit/069230e)

<br>

### 3. 불필요한 useEffect 제거

- 커스텀 훅 `useValidateForm` 내부의 불필요한 useEffect 제거 [(6d96dac)](https://github.com/woowacourse/react-payments/commit/6d96dac)
  - useEffect를 사용하지 않아도, 로직들이 상태로 관리하는 데이터를 참조하기 때문에 데이터가 변경될 때 자동으로 업데이트 되므로 제거함

- 커스텀 훅 `useRender`의 불필요한 useEffect 제거 [(2486c59)](https://github.com/woowacourse/react-payments/commit/2486c59)
  - useEffect가 참조하는 order는 상태로 관리되고 있으며, useEffect 내부에서 order의 상태를 변경하고 있으므로 불필요한 렌더링이 발생함

<br>

### 4. 뒤로가기 시, 이전 form 데이터가 남아있는 문제 개선

- 뒤로가기로 카드 정보 입력 페이지로 돌아가면 기존의 데이터가 불완전하게 남아있는 문제를 개선함 [(5dd25e5)](https://github.com/woowacourse/react-payments/commit/5dd25e5)
  - 기존의 코드는 등록 완료 페이지에서 확인 버튼을 클릭할 때, form 초기화 이벤트가 발생하도록 설정되어 있었음
  - 그러나 바로가기를 클릭하면 form 초기화를 하지 않고 페이지에 접근하기 때문에 기존의 데이터가 남아 있는 문제가 발생함
  - 따라서 완료 페이지로 넘어갈 때 form을 초기화하여, 뒤로가기로 페이지에 접근해도 초기화 된 form을 렌더링하도록 개선

<br>

### 5. UI/UX 및 스타일 개선
- Input 컴포넌트에 `pattern` 속성 추가 [(957695e)](https://github.com/woowacourse/react-payments/commit/957695e)
  - 기존에 예외 처리의 기준이 되었던 `minLength` 속성은 submit 이벤트가 발생될 때 검증되는 값임. 따라서 실시간으로 변경되는 사용자의 입력값을 검증하기에는 정확도가 떨어지고, 실제로도 `minLength`를 충족하지 못했음에도 유효하다고 판별하는 버그가 있었음. 따라서 실시간으로 유효성 검증이 가능한 `pattern` 속성 추가

- 피그마 요구사항에 맞게 카드 정보 입력 UI가 아래에서 위로 누적되도록 수정 [(91f6abf)](https://github.com/woowacourse/react-payments/commit/91f6abf)

- 모바일 환경에서 확인 버튼이 표시될 때 레이아웃이 변경되는 문제 개선 [(229e8e6)](https://github.com/woowacourse/react-payments/commit/229e8e6)

<br>
<br>

## 기타 변경사항

- 논리 파악에 혼란을 주기 쉬운 예외 처리 문법 개선 [(378e0f0)](https://github.com/woowacourse/react-payments/commit/378e0f0)
- 불필요한 중복 Context Provider 제거 [(c933a32)](https://github.com/woowacourse/react-payments/commit/c933a32)
- React가 list items 형태로 렌더링되는 Input들을 인식할 수 있도록 key 부여 [(c06b52b)](https://github.com/woowacourse/react-payments/commit/c06b52b)
- 새롭게 렌더링되는 카드 정보 UI에 자동으로 포커싱이 가도록 수정 [(773b4fd)](https://github.com/woowacourse/react-payments/commit/773b4fd)
- 예외처리 로직에서 precise 직접 호출을 validate 함수 사용으로 대체 [(6b5c0f3)](https://github.com/woowacourse/react-payments/commit/6b5c0f3)
- 변수명의 통일성을 고려하여 renderList를 renderItems로 수정 [(bf9b794)](https://github.com/woowacourse/react-payments/commit/bf9b794)

<br>
<br>

## 추가 질문

### 1. 카드 유효기간 유효성 검사 로직의 개선안

현재의 유효기간 검증 로직에서 UX 측면에서 어떤 점을 개선할 수 있을지 도밥의 의견이 궁금합니다.

이번 리팩토링에서 가장 어려웠던 작업은 카드 유효기간의 유효성 검사 로직 개선하는 것이었습니다. 특히 onBlur 이벤트를 활용해 입력이 완전히 끝난 뒤에만 검사를 실행할지 고민했는데, 이 방식은 사용자가 마지막 필드를 채운 뒤, 포커스를 직접 해제해야만 검사가 실행되는 제약이 있었습니다. 결국 그대로 onChange 핸들러를 통해 실시간으로 검증하기로 결정했습니다.

또한 사용자가 월(month) 입력을 완성하지 않은 상태에서 연(year)을 입력하거나, 잘못된 월 값이 입력된 채로 연도를 입력하는 등 다양한 예외 상황을 모두 다루기가 까다로웠습니다. 이러한 복합적인 경우를 고려한 자연스러운 예외 처리 로직을 설계하는 데에도 많은 고민을 했지만, 생각대로 동작이 잘 구현되지 않았습니다.

현재의 코드에서 UX 혹은 복잡한 로직을 보다 직관적으로 바꾸기 위해, 어떤 점을 고민해 보면 좋을지 도밥에게 여쭙고 싶습니다..!

Creative-Lee reviewed (approved):
안녕하세요 제나 😄
미션 물면서 진행하느라 힘드시죠~! 늦은 새벽에 정말 고생하셨어요👍 

> 1. 카드 유효기간 유효성 검사 로직의 개선안
특히 onBlur 이벤트를 활용해 입력이 완전히 끝난 뒤에만 검사를 실행할지 고민했는데, 이 방식은 사용자가 마지막 필드를 채운 뒤, 포커스를 직접 해제해야만 검사가 실행되는 제약이 있었습니다. 결국 그대로 onChange 핸들러를 통해 실시간으로 검증하기로 결정했습니다.

onBlur로 피드백 시점을 변경? 해보려고 했으나 제약(input focus를 해제해야만 피드백)이 있었고 이게 맘에 들지 않아(제나가 의도한 UX가 아니라) onChange로 변경한것으로 이해했어요.
사용자의 포커스가 어떤 input인지를 떠나, 실시간으로 값을 입력하는 단계에서 에러 피드백을 주고자 했다면 onChange 핸들러가 적합했을 거에요. 
제나가 구현하고자 했던 UX가 정확히 어떤것인지 몰라서 답변이 쉽지않네요ㅎㅎ..

> 또한 사용자가 월(month) 입력을 완성하지 않은 상태에서 연(year)을 입력하거나, 잘못된 월 값이 입력된 채로 연도를 입력하는 등 다양한 예외 상황을 모두 다루기가 까다로웠습니다. 이러한 복합적인 경우를 고려한 자연스러운 예외 처리 로직을 설계하는 데에도 많은 고민을 했지만, 생각대로 동작이 잘 구현되지 않았습니다. 현재의 코드에서 UX 혹은 복잡한 로직을 보다 직관적으로 바꾸기 위해, 어떤 점을 고민해 보면 좋을지 도밥에게 여쭙고 싶습니다..!

코드를 보고 딱 집어드리기는 어려워서, 조금 전체적인 방향으로 코멘트를 드리면...
다양한 유즈 케이스로 인해 혼란이 생길때는 제일 일반적인 케이스부터 처리하며 확장해 나가는 방식이 좋은 것 같습니다. 
또한 전체 form의 동작과 관련한 내용은 잠시 컨텍스트에서 제외시켜두고, 
하나의 input에서 어떤 동작을 기대하는지, 이 input의 에러피드백은 어떤 시점에 어떤 조건으로 발생하는지 등을 챙겨 구현하고 조합해서 마무리하는 흐름으로 생각해봐도 좋을 것 같아요.
(e.g. 월, 년 입력 순서나 유효성 관련해서 복잡함을 겪은것 같은데, 이걸 우선 '유효한 월/년 을 입력해주세요' 처럼 포괄적인 처리로 구현하고 세분화 해볼수도 있겠습니다.)

여담으로... react에서 form을 다루는것은 원래 악명높고 까다롭습니다(저 역시 매일 씨름해요😅)
그래도 지금 이렇게 헤딩하는 시간들이 추후에 도움이 많이 될거라고 생각합니다.
(form을 쉽게 다룰 수 있게 해주는 라이브러리등을 사용할때도 더 잘 활용할 수 있을것이고요)

---

useState, useEffect 관련 코멘트 남겨드리며 강조했었는데, 
이 부분에 대해서 잘 학습해주신것이 코드에서 보여서 좋네요 ㅎㅎ 잘 반영해주셨습니다~!
(불필요한 state와 effect가 많이 줄어서 코드 흐름도 훨씬 깔끔해졌네요)
react 첫 미션에서 얻어야할 것들은 잘 챙기셨다고 생각합니다👍
 
이번 미션 정말 고생많으셨고 앞으로도 화이팅입니다😄 
다음에 또 봐요 제나👋
